lib/Platform/Intl/PlatformIntlApple.mm

```diff
@@ -68,12 +208,69 @@
   return canonicalizeLocaleList(runtime, locales);
 }
 
+// Implementer note: This method corresponds roughly to
+// https://tc39.es/ecma402/#sup-string.prototype.tolocalelowercase
 vm::CallResult<std::u16string> toLocaleLowerCase(
     vm::Runtime *runtime,
     const std::vector<std::u16string> &locales,
     const std::u16string &str) {
-  return std::u16string(u"lowered");
+  NSString *nsStr = u16StringToNSString(str);
+
+  // 3. Let requestedLocales be ? CanonicalizeLocaleList(locales).
+  vm::CallResult<std::vector<std::u16string>> requestedLocales =
+      getCanonicalLocales(runtime, locales);
+  // getcano doesnt throw so should this be removed?
+  if (LLVM_UNLIKELY(requestedLocales == llvh::ExecutionStatus::EXCEPTION)) {
+    return llvh::ExecutionStatus::EXCEPTION;
+  }
+
+  // 4. If requestedLocales is not an empty List, then
+  std::u16string requestedLocale;
+  if (!requestedLocales->empty()) {
+    // a. Let requestedLocale be requestedLocales[0].
+    std::vector<std::u16string> val = requestedLocales.getValue();
+    requestedLocale = val[0];
+  } else { // 5. Else,
+    // a. Let requestedLocale be DefaultLocale().
+    return nsStringToU16String(
+        [nsStr lowercaseStringWithLocale:[NSLocale currentLocale]]);
+  }
+  // 6. Let noExtensionsLocale be the String value that is requestedLocale with
+  // any Unicode locale extension sequences (6.2.1) removed.
+  std::u16string noExtensionsLocale = toNoExtensionsLocale(requestedLocale);
+
+  // 7. Let availableLocales be a List with language tags that includes the
+  // languages for which the Unicode Character Database contains language
+  // sensitive case mappings. Implementations may add additional language tags
+  // if they support case mapping for additional locales.
+  // 8. Let locale be BestAvailableLocale(availableLocales, noExtensionsLocale).
+  // Convert to C++ array for bestAvailableLocale function
+  static std::vector<std::u16string> availableLocalesVector = getAvailableLocalesVector();
+  llvh::Optional<std::u16string> locale =
+      bestAvailableLocale(availableLocalesVector, noExtensionsLocale);
+  // 9. If locale is undefined, let locale be "und".
+  if (!locale.hasValue()) {
+    locale = u"und";
+  }
```

Comment:
All of this looks like it will be reusable, at the very least in `toLocaleUpperCase`. Let's factor it out into a function that takes a locale list and returns the locale name as a `u16string`. This does mean that we'll need to extract the default locale as a string and then reconstruct the `NSLocale`, but it keeps the option to potentially reuse this code on other platforms.